Add full volume control to SCC

pweil- opened the pull request:
Add full volume control to SCC.  In upstream PSP we have proposed using a Volumes slice on the SCC to control volumes.  This introduces that attribute to the SCC

Fixes https://github.com/openshift/origin/issues/7630

Important: 

1.  ~~the comment on PSP was that if the slice was empty you are allowing all volumes.  This is not the case in this PR.  I think we can argue against it upstream if need be.  Nothing has been implemented yet to enforce it in kube.~~ updated defaulting to consider `nil` as allowing all volumes, empty allows none based on feedback
1.  please take note of how the interaction with `Allow*` fields works in conversions and defaulters

Open Questions

1.  ~~Can we retcon the addition of `AllowEmptyDir` in the SCC?  It was added and tagged in 1.1.3 but nothing was noted in the release for it.  I'd like to just rip it out if possible.  For now I've included the conversion code for it.~~ removed

- [x] types
- [x] conversions, generation, and defaulting
- [x] enforcement in the provider
- [x] bootstrap settings

@smarterclayton @liggitt @pmorie @childsb @eparis

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1/conversion_generated.go:
```diff
@@ -2958,8 +2958,22 @@ func autoConvert_api_SecurityContextConstraints_To_v1_SecurityContextConstraints
+	}
+
+	// set the Allow* fields based on the existence in the volume slice
+	volumeStringSet := fsTypeToStringSet(out.Volumes)
```
move these up into the loop through the volume types.. a switch inside the loop with cases for hostpath/emptydir saves you a stringset construction. @smarterclayton, can we drop `AllowEmptyDirVolumePlugin`?

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1/defaults.go:
```diff
@@ -323,4 +324,70 @@ func defaultSecurityContextConstraints(scc *SecurityContextConstraints) {
+}
+
+// defaultSecurityContextConstraintsVolumes defaults the volume slice of the SCC.
+// The volumes slice will take precedence in settings if it is set.
```
this is going to be more complicated than you'd like... 
* old go clients will drop the volumes slice (so it will be nil), and set the booleans
* old naive clients will just touch the fields they know about (the booleans), and leave the other fields alone (like a responsible client), so the booleans have to take precedence over the volumes slice
* new go clients will serialize from the volumes slice internally to a correct external volumes slice and correct external booleans, so yay!
* new naive clients will have to set both the volumes slice and the booleans (or *possibly* omit the booleans and really hope they're talking to a *new* v1 server, which is iffy, and only works if our booleans are pointers and we can tell they were omitted)

liggitt replied:
the takeaway is that the booleans always win. clients will have to set the booleans AND the volumes slice (or maybe omit the booleans if we make them pointers, but I think that breaks clients talking to older v1 servers)

smarterclayton replied:
It was amusing watching you work through that :)

smarterclayton replied:
It's easy to forget how bad it is.

liggitt replied:
The. Booleans. Always. Win.

smarterclayton replied:
That's True.

pweil- commented:
@liggitt - updated defaulting logic to always use the bool values in the slice and updated test cases to reflect.  PTAL at that portion.

pweil- commented:
Updated with enforcement pieces and bootstrap scc updates

pweil- commented:
ready for more eyeballs :eyes:

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/types.go:
```diff
@@ -2507,6 +2505,30 @@ type SecurityContextConstraints struct {
 }
 
+// FS Type gives strong typing to different file systems that are used by volumes.
+type FSType string
```
need a reflective test to detect drift if a new field is added to `VolumeSource`

liggitt replied:
`VolumeType`?

pweil- replied:
This also matches the upstream approved type so I'd like to keep it the same

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/provider.go:
```diff
@@ -353,3 +352,69 @@ func createSupplementalGroupStrategy(opts *api.SupplementalGroupsStrategyOptions
+}
+
+func getVolumeFSType(v api.Volume) api.FSType{
+	if v.HostPath != nil {
```
```
switch {
case v.HostPath != nil:
  return api.FSTypeHostPath
...
}
```

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/provider.go:
```diff
@@ -353,3 +352,69 @@ func createSupplementalGroupStrategy(opts *api.SupplementalGroupsStrategyOptions
+	if v.AzureFile != nil {
+		return api.FSTypeAzureFile
+	}
+	return api.FSType("unknown type")
```
I'd expect an error here

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/types.go:
```diff
@@ -2507,6 +2505,30 @@ type SecurityContextConstraints struct {
+	FSTypeCephFS                FSType = "cephFS"
+	FSTypeDownwardAPI           FSType = "downwardAPI"
+	FSTypeFC                    FSType = "fc"
+)
```
I think I'd like a `FSTypeAll = "*"` we can grant to the highest level sccs

pweil- replied:
+1

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/types.go:
```diff
@@ -2480,10 +2480,8 @@ type SecurityContextConstraints struct {
-	// AllowEmptyDirVolumePlugin determines if the policy allow containers to use the EmptyDir volume plugin
-	AllowEmptyDirVolumePlugin bool
+	// Volumes is a white list of allowed volume plugins.
+	Volumes []FSType
```
`AllowedVolumeTypes`? `AllowedVolumeSources`?

pweil- replied:
The upstream field is Volumes.  We got away from the Allow/Deny wording.

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1/conversion_generated.go:
```diff
@@ -2958,8 +2958,31 @@ func autoConvert_api_SecurityContextConstraints_To_v1_SecurityContextConstraints
+	if in.Volumes != nil {
+		out.Volumes = make([]FSType, len(in.Volumes))
+		for i, v := range in.Volumes {
+			out.Volumes[i] = FSType(in.Volumes[i])
```
shouldn't need a typecast

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1/conversion_generated.go:
```diff
@@ -6278,10 +6301,18 @@ func autoConvert_v1_SecurityContextConstraints_To_api_SecurityContextConstraints
+	if in.Volumes != nil {
+		out.Volumes = make([]api.FSType, len(in.Volumes))
+		for i := range in.Volumes {
+			out.Volumes[i] = api.FSType(in.Volumes[i])
```
no typecast?

pweil- commented on pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go:
```diff
@@ -63,13 +63,12 @@ func GetBootstrapSecurityContextConstraints(sccNameToAdditionalGroups map[string
-			AllowHostPID:              true,
-			AllowHostIPC:              true,
+			AllowPrivilegedContainer: true,
+			Volumes:                  []kapi.FSType{kapi.FSTypeHostPath, kapi.FSTypeEmptyDir, kapi.FSTypeSecret},
```
No, this was just to get all the mechanics worked out by replacing the existing bools.  We can open this list up

pweil- commented:
Updated with `*` wildcard as `FSTypeAll`, added new reflective test for drift.  Also added a reflective test that will test every volumesource with a deny, accept via explicit add to SCC, and accept via `FSTypeAll`.

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/types.go:
```diff
@@ -2507,6 +2505,32 @@ type SecurityContextConstraints struct {
+	FSTypeDownwardAPI           FSType = "downwardAPI"
+	FSTypeFC                    FSType = "fc"
+
+	FSTypeAll				    FSType = "*"
```
gofmt should complain about this

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1beta3/defaults.go:
```diff
@@ -237,4 +238,46 @@ func defaultSecurityContextConstraints(scc *SecurityContextConstraints) {
+
+// defaultSecurityContextConstraintsVolumes defaults the volume slice of the SCC.
+// In order to support old clients the boolean fields will always take precedence.
+func defaultSecurityContextConstraintsVolumes(scc *SecurityContextConstraints) {
```
this depends on being called after other defaulting bits... don't split it into a function that could be called incorrectly, just keep these bits in the defaultSecurityContextConstraints func

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1beta3/types.go:
```diff
@@ -2130,6 +2132,32 @@ type SecurityContextConstraints struct {
+	FSTypeDownwardAPI           FSType = "downwardAPI"
+	FSTypeFC                    FSType = "fc"
+
+	FSTypeAll				    FSType = "*"
```
gofmt

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/describe.go:
```diff
@@ -452,7 +452,7 @@ func describeSecurityContextConstraints(scc *api.SecurityContextConstraints) (st
 		fmt.Fprintf(out, "  Required Drop Capabilities:\t%s\n", capsToString(scc.RequiredDropCapabilities))
 		fmt.Fprintf(out, "  Allowed Capabilities:\t%s\n", capsToString(scc.AllowedCapabilities))
-		fmt.Fprintf(out, "  Allow Host Dir Volumes:\t%t\n", scc.AllowHostDirVolumePlugin)
+		fmt.Fprintf(out, "  Allowed Volumes:\t%s\n", fsTypeToStringSet(scc.Volumes))
```
volume types?

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/describe.go:
```diff
@@ -507,6 +507,18 @@ func stringOrNone(s string) string {
 	return "<none>"
 }
 
+func fsTypeToStringSet(volumes []api.FSType) string {
```
fsTypeToString

liggitt replied:
can just do 
```
strVolumes := []string{}
for _, v := range volumes {
	strVolumes = append(strVolumes, string(v))
}
return stringOrNone(strings.Join(strVolumes, ","))
```

smarterclayton commented:
We need to release note in 1.1.4 that users upgrading should run SCC reconcile.  Otherwise LGTM, do we have other comments?

liggitt commented:
LGTM, do we want to pull in https://github.com/openshift/origin/pull/7776 first to avoid the emptydir churn?

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1/defaults.go:
```diff
@@ -323,4 +324,42 @@ func defaultSecurityContextConstraints(scc *SecurityContextConstraints) {
+	return volumes
+}
+
+func fsTypeToStringSet(volumes []FSType) sets.String {
```
actually, this is asymmetrical... a nil slice would convert to an empty (but non-nil) string set

liggitt replied:
this becomes important when we remove the emptyDir bool... a nil volumes slice has to default to FSTypeEmptyDir for compatibility, but an empty slice can really mean no volume types are allowed

liggitt replied:
still asymmetrical

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1/defaults.go:
```diff
@@ -323,4 +324,42 @@ func defaultSecurityContextConstraints(scc *SecurityContextConstraints) {
+}
+
+func stringSetToFSType(set sets.String) []FSType {
+	if set.Len() == 0 {
```
and an empty-but-not-nil string set converts to a nil FSType slice

smarterclayton commented:
Probably.  You are doing the rebase so you have merge control on UPSTREAM.

liggitt commented:
then revert emptydir first, and fix up the nil/empty slice defaulting logic here to always default a nil volumes slice to one containing FSEmptyDir

pweil- commented:
Agree, getting the emptydir reverted first will reduce some of this code.  I'll fix up the slice pieces you mentioned.  Thanks for the feedback

liggitt commented:
* Needs rebase
* `FSTypeAll` should convert to AllowHostDir: true
* Should a nil slice default to `FSTypeAll`, since that is what it is currently allowing? A nil slice + `AllowHostDir: false` would mean `* - HostDir`, which is unfortunate...

pweil- commented:
@liggitt I reworked the defaulting and conversions in the last commit to try and capture the effects of defaulting FSTypeAll for empty slices.  PTAL at the last commit.

smarterclayton commented:
You need to allow downwardAPI.  What about config map?

On Tue, Mar 8, 2016 at 10:28 AM, Paul Weil <notifications@github.com> wrote:

> @liggitt <https://github.com/liggitt> I reworked the defaulting and
> conversions in the last commit to try and capture the effects of defaulting
> FSTypeAll for empty slices. PTAL at the last commit.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7641#issuecomment-193850305>.
>

pweil- commented:
> You need to allow downwardAPI.  What about config map?

downward api is covered.  Config map is actually not represented by a field in `VolumeSource` yet.  It should come in in the next rebase.  I can make the FSType changes for it now though.

smarterclayton commented on pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go:
```diff
@@ -147,11 +148,11 @@ func GetBootstrapSecurityContextConstraints(sccNameToAdditionalGroups map[string
-			AllowHostPorts:           true,
-			AllowHostPID:             true,
-			AllowHostIPC:             true,
+			Volumes:          []kapi.FSType{kapi.FSTypeHostPath, kapi.FSTypeEmptyDir, kapi.FSTypeSecret},
```
DownwardAPI

pweil- replied:
oh there...heh...adding.

smarterclayton commented on pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go:
```diff
@@ -118,7 +119,7 @@ func GetBootstrapSecurityContextConstraints(sccNameToAdditionalGroups map[string
 				},
 			},
-			AllowHostDirVolumePlugin: true,
+			Volumes: []kapi.FSType{kapi.FSTypeHostPath, kapi.FSTypeEmptyDir, kapi.FSTypeSecret},
```
DownwardAPI

smarterclayton commented on pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go:
```diff
@@ -179,6 +180,7 @@ func GetBootstrapSecurityContextConstraints(sccNameToAdditionalGroups map[string
 					DescriptionAnnotation: SecurityContextConstraintRestrictedDesc,
 				},
 			},
+			Volumes: []kapi.FSType{kapi.FSTypeEmptyDir, kapi.FSTypeSecret},
```
DownwardAPI

smarterclayton commented on pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go:
```diff
@@ -208,6 +210,7 @@ func GetBootstrapSecurityContextConstraints(sccNameToAdditionalGroups map[string
 					DescriptionAnnotation: SecurityContextConstraintsAnyUIDDesc,
 				},
 			},
+			Volumes: []kapi.FSType{kapi.FSTypeEmptyDir, kapi.FSTypeSecret},
```
DownwardAPI

smarterclayton commented on pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go:
```diff
@@ -238,6 +241,7 @@ func GetBootstrapSecurityContextConstraints(sccNameToAdditionalGroups map[string
 			},
 			AllowHostNetwork: true,
 			AllowHostPorts:   true,
+			Volumes:          []kapi.FSType{kapi.FSTypeEmptyDir, kapi.FSTypeSecret},
```
DownwardAPI

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1/defaults.go:
```diff
@@ -317,4 +319,55 @@ func defaultSecurityContextConstraints(scc *SecurityContextConstraints) {
+	defaultAllowedVolumes := fsTypeToStringSet(scc.Volumes)
+
+	// assume an empty volume slice is allowing everything for backwards compatibility
+	if defaultAllowedVolumes.Len() == 0 {
```
no, `nil` means everything. empty means nothing.

liggitt replied:
that sounded more philosophical than I meant

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1/defaults.go:
```diff
@@ -317,4 +319,55 @@ func defaultSecurityContextConstraints(scc *SecurityContextConstraints) {
+	// we should only default all volumes if the SCC came in with FSTypeAll or we defaulted it
+	// otherwise we should only change the volumes slice to ensure that it does not conflict with
+	// the AllowHostDirVolumePlugin setting
+	shouldDefaultAllVolumes := defaultAllowedVolumes.Has(string(FSTypeAll))
```
move inside the `else` case to keep the logic closer to where it is used

liggitt replied:
all the same comments apply to v1beta3, I think

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1/defaults_test.go:
```diff
@@ -723,3 +724,77 @@ func TestDefaultSecurityContextConstraints(t *testing.T) {
+			},
+			expectedVolumes: []versioned.FSType{versioned.FSTypeAll},
+			expectedHostDir: true,
+		},
```
add the following tests:
* empty volumes slice (`[]`), allowHostDir:false, should result in empty slice
* empty volumes slice (`[]`), allowHostDir:true, should result in hostdir slice

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1/types.go:
```diff
@@ -2973,6 +2975,32 @@ type SecurityContextConstraints struct {
+	FSTypeCephFS                FSType = "cephFS"
+	FSTypeDownwardAPI           FSType = "downwardAPI"
+	FSTypeFC                    FSType = "fc"
+	FSTypeConfigMap			    FSType = "configMap"
```
indent?

pweil- replied:
WTH is wrong with my IDE... :-\

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/util/util.go:
```diff
@@ -0,0 +1,120 @@
+}
+
+// getVolumeFSType gets the FSType for a volume.
+func GetVolumeFSType(v api.Volume) (api.FSType, error) {
```
does volume validation ensure we'll never have more than one of these set (it should)? don't want to short circuit here and allow a pod through when it specifies two volume types in a single volume

pweil- replied:
yes, volume validation will ensure that a `volumeSource` may not specify more than one type of volume

liggitt commented on pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go:
```diff
@@ -91,6 +91,7 @@ func GetBootstrapSecurityContextConstraints(sccNameToAdditionalGroups map[string
 					DescriptionAnnotation: SecurityContextConstraintNonRootDesc,
 				},
 			},
+			Volumes: []kapi.FSType{kapi.FSTypeEmptyDir, kapi.FSTypeSecret},
```
FSTypeConfigMap, FSTypeDownwardAPI?

liggitt commented on pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go:
```diff
@@ -118,7 +119,7 @@ func GetBootstrapSecurityContextConstraints(sccNameToAdditionalGroups map[string
 				},
 			},
-			AllowHostDirVolumePlugin: true,
+			Volumes: []kapi.FSType{kapi.FSTypeHostPath, kapi.FSTypeEmptyDir, kapi.FSTypeSecret, kapi.FSTypeDownwardAPI},
```
FSTypeConfigMap?

liggitt commented on pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go:
```diff
@@ -147,11 +148,11 @@ func GetBootstrapSecurityContextConstraints(sccNameToAdditionalGroups map[string
-			AllowHostPorts:           true,
-			AllowHostPID:             true,
-			AllowHostIPC:             true,
+			Volumes:          []kapi.FSType{kapi.FSTypeHostPath, kapi.FSTypeEmptyDir, kapi.FSTypeSecret, kapi.FSTypeDownwardAPI},
```
FSTypeConfigMap?

liggitt commented on pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go:
```diff
@@ -179,6 +180,7 @@ func GetBootstrapSecurityContextConstraints(sccNameToAdditionalGroups map[string
 					DescriptionAnnotation: SecurityContextConstraintRestrictedDesc,
 				},
 			},
+			Volumes: []kapi.FSType{kapi.FSTypeEmptyDir, kapi.FSTypeSecret, kapi.FSTypeDownwardAPI},
```
FSTypeConfigMap?

liggitt commented on pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go:
```diff
@@ -208,6 +210,7 @@ func GetBootstrapSecurityContextConstraints(sccNameToAdditionalGroups map[string
 					DescriptionAnnotation: SecurityContextConstraintsAnyUIDDesc,
 				},
 			},
+			Volumes: []kapi.FSType{kapi.FSTypeEmptyDir, kapi.FSTypeSecret, kapi.FSTypeDownwardAPI},
```
FSTypeConfigMap?

liggitt commented on pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go:
```diff
@@ -238,6 +241,7 @@ func GetBootstrapSecurityContextConstraints(sccNameToAdditionalGroups map[string
 			},
 			AllowHostNetwork: true,
 			AllowHostPorts:   true,
+			Volumes:          []kapi.FSType{kapi.FSTypeEmptyDir, kapi.FSTypeSecret, kapi.FSTypeDownwardAPI},
```
FSTypeConfigMap?

liggitt commented on pkg/security/admission/byrestrictions.go:
```diff
@@ -50,3 +50,13 @@ func pointValue(constraint *kapi.SecurityContextConstraints) int {
 }
+
+// allowsHostPathVolume checks to see if the FSTypeHostPath is in the Volumes slice.
+func allowsHostPathVolume(scc *kapi.SecurityContextConstraints) bool {
```
probably any volume type other than emptydir/secret/configmap/downwardapi is worth bumping priority

pweil- commented:
updated for latest feedback

[test]

pweil- commented:
ready for another look

smarterclayton commented:
No more comments from me.

kargakis commented on pkg/security/admission/byrestrictions.go:
```diff
@@ -50,3 +48,37 @@ func pointValue(constraint *kapi.SecurityContextConstraints) int {
+// Allowing a host volume wil return a score of 10.  Allowance of anything other
+// than kapi.FSTypeSecret, kapi.FSTypeConfigMap, kapi.FSTypeConfigMap, kapi.FSTypeDownwardAPI
+// will result in a score of 5.  If the SCC only allows kapi.FSTypeSecret, kapi.FSTypeConfigMap,
+// kapi.FSTypeConfigMap, kapi.FSTypeDownwardAPI it will have a score of 0.
```
s/kapi.FSTypeConfigMap/kapi.FSTypeEmptyDir/

pmorie replied:
Also, I think git repo should be in this list...

pmorie commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/types.go:
```diff
@@ -2505,8 +2505,8 @@ type SecurityContextConstraints struct {
 	AllowedCapabilities []Capability
-	// AllowHostDirVolumePlugin determines if the policy allow containers to use the HostDir volume plugin
-	AllowHostDirVolumePlugin bool
+	// Volumes is a white list of allowed volume plugins.
```
I think it would be helpful to define the format of this format, since as far as I know the constants aren't rippled out to the user-doc for this.

pmorie replied:
I do realize this is probably similar to the doc in the upstream patch :-/

pmorie commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/resource_printer.go:
```diff
@@ -420,7 +420,7 @@ var withNamespacePrefixColumns = []string{"NAMESPACE"} // TODO(erictune): print
 var configMapColumns = []string{"NAME", "DATA", "AGE"}
 var podSecurityPolicyColumns = []string{"NAME", "PRIV", "CAPS", "VOLUMEPLUGINS", "SELINUX", "RUNASUSER"}
-var securityContextConstraintsColumns = []string{"NAME", "PRIV", "CAPS", "HOSTDIR", "SELINUX", "RUNASUSER", "FSGROUP", "SUPGROUP", "PRIORITY"}
+var securityContextConstraintsColumns = []string{"NAME", "PRIV", "CAPS", "VOLUMES", "SELINUX", "RUNASUSER", "FSGROUP", "SUPGROUP", "PRIORITY"}
```
how do you feel about having this at the end, because it might be super-huge?

pmorie commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/provider.go:
```diff
@@ -243,10 +244,19 @@ func (s *simpleProvider) ValidateContainerSecurityContext(pod *api.Pod, containe
+			if !allowedVolumes.Has(string(fsType)) {
+				allErrs = append(allErrs, field.Invalid(
+					fldPath.Child("volumes").Index(i), string(fsType),
+					fmt.Sprintf("%s volumes are not allowed to be used", string(fsType))))
```
nit: i _think_ if you do %v here, you can drop the string conversion.

pmorie commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/util/util.go:
```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
```
wrong year

pmorie commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/util/util_test.go:
```diff
@@ -0,0 +1,47 @@
+	"k8s.io/kubernetes/pkg/api"
+)
+
+func TestVolumeSourceFSTypeDrift(t *testing.T) {
```
I think a comment would be beneficial here, this test is really doing two things:

1.  testing the integrity of `GetAllFSTypesAsSet`
2.  testing that for all the fields of volume source, an fstype pops out

pmorie commented on pkg/cmd/admin/policy/reconcile_sccs_test.go:
```diff
@@ -15,7 +15,7 @@ func TestComputeDefinitions(t *testing.T) {
 
 	diffHostDir := goodSCC()
-	diffHostDir.AllowHostDirVolumePlugin = true
+	diffHostDir.Volumes = []kapi.FSType{kapi.FSTypeHostPath}
```
do you need to retain the old test?

pweil- replied:
internal no longer has the Allow field so it must be tested through the Volumes slice.

pmorie commented on pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go:
```diff
@@ -147,11 +148,11 @@ func GetBootstrapSecurityContextConstraints(sccNameToAdditionalGroups map[string
-			AllowHostPorts:           true,
-			AllowHostPID:             true,
-			AllowHostIPC:             true,
+			Volumes:          []kapi.FSType{kapi.FSTypeHostPath, kapi.FSTypeEmptyDir, kapi.FSTypeSecret, kapi.FSTypeDownwardAPI, kapi.FSTypeConfigMap},
```
what about git repo here?

pmorie replied:
I think it will break people unless git repo is in the lists added in this file

pweil- replied:
@smarterclayton @liggitt - thoughts on gitrepo being included as a default?

smarterclayton replied:
We haven't tested with it, can be left out for now.  Open a low severity issue.

pweil- replied:
https://github.com/openshift/origin/issues/8045

pweil- commented:
@pmorie - feedback updates made with the exception of the gitrepo piece.  Squashed into the existing feedback commits.  PTAL

pweil- commented:
rebased and squashed

liggitt commented:
Same for this. We want to make sure the generate scripts upstream keep working in the fork

pweil- commented:
Will do, been stuck in registry land today.  Since there are two SCC PRs open and this is the more complicated one I would prefer to land the read only root fs PR first then I'll rebase this on top of it.

liggitt commented:
Sure

pweil- commented:
Well, the drift test is working well...caught the config map piece that came in.  will fix

```
--- FAIL: TestVolumeSourceFSTypeDrift (0.00s)
	util_test.go:44: error getting fstype for field ConfigMap.  This likely means that drift has occured between FSType and VolumeSource.  Please update the api and getVolumeFSType
```

smarterclayton commented:
We should add a drift test for our drift tests. Then a drift test for
the drift test of drift tests.

pweil- commented:
`hack/update-generated-drifttest-drifttest-drifttests.sh`

pweil- commented:
https://github.com/openshift/kubernetes/pull/13

pweil- commented:
re[test]

pweil- commented:
re[test] - ready for more reviews

pweil- commented:
green tests.  @liggitt want to give this a final pass or can we merge away?

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1beta3/conversion_generated.go:
```diff
@@ -2053,7 +2053,22 @@ func convert_api_SecurityContextConstraints_To_v1beta3_SecurityContextConstraint
+		for i, v := range in.Volumes {
+			out.Volumes[i] = FSType(in.Volumes[i])
+
+			// set the Allow* fields based on the existence in the volume slice
```
custom conversions shouldn't be in conversion_generated...

pweil- replied:
ah yes, I just need to remove this completely if the v1beta3 api won't have this field.  It didn't complain because os/kube doesn't generate for that version.

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1beta3/types.go:
```diff
@@ -2105,6 +2105,9 @@ type SecurityContextConstraints struct {
 	AllowHostDirVolumePlugin bool `json:"allowHostDirVolumePlugin" description:"allow the use of the host dir volume plugin"`
+	// Volumes is a white list of allowed volume plugins.  FSType corresponds directly with the field names
+	// of a VolumeSource (azureFile, configMap, emptyDir).  To allow all volumes you may use '*'.
+	Volumes []FSType `json:"volumes"`
```
don't need the new field... we'll never round trip to v1beta3

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/v1beta3/defaults.go:
```diff
@@ -231,4 +233,58 @@ func defaultSecurityContextConstraints(scc *SecurityContextConstraints) {
+	// In order to support old clients the boolean fields will always take precedence.
+	defaultAllowedVolumes := fsTypeToStringSet(scc.Volumes)
+
+	// assume a nil volume slice is allowing everything for backwards compatibility
```
I think this should all be a custom conversion... that keeps us from needing a new field, new deep copy, etc

liggitt commented:
split v1beta3 bits into a commit

pweil- commented:
@liggitt field has been removed from v1beta3, commit split, and custom conversions are registered to set internal volumes based on the v1beta3 AllowHostDir field

liggitt commented:
LGTM

smarterclayton commented:
[merge]

On Mon, Mar 28, 2016 at 11:15 AM, Jordan Liggitt <notifications@github.com>
wrote:

> LGTM
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7641#issuecomment-202432956>
>
